render/vulkan/vulkan.c

```diff
@@ -0,0 +1,596 @@
+				phdev_props.deviceName);
+			return phdev;
+		}
+	}
+
+	return VK_NULL_HANDLE;
+}
+
+struct wlr_vk_device *vulkan_device_create(struct wlr_vk_instance *ini,
+		VkPhysicalDevice phdev, size_t ext_count, const char **exts) {
+	VkResult res;
+
+	// check for extensions
+	uint32_t avail_extc = 0;
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, NULL);
+	if (res != VK_SUCCESS || avail_extc == 0) {
+		wlr_vk_error("Could not enumerate device extensions (1)", res);
+		return NULL;
+	}
+
+	VkExtensionProperties avail_ext_props[avail_extc + 1];
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, avail_ext_props);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("Could not enumerate device extensions (2)", res);
+		return NULL;
+	}
+
+	for (size_t j = 0; j < avail_extc; ++j) {
+		wlr_log(WLR_DEBUG, "Vulkan device extension %s v%"PRIu32,
+			avail_ext_props[j].extensionName, avail_ext_props[j].specVersion);
+	}
+
+	// create device
+	struct wlr_vk_device *dev = calloc(1, sizeof(*dev));
+	if (!dev) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		return NULL;
+	}
+
+	dev->phdev = phdev;
+	dev->instance = ini;
+	dev->drm_fd = -1;
+	dev->extensions = calloc(16 + ext_count, sizeof(*ini->extensions));
+	if (!dev->extensions) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		goto error;
+	}
+
+	// find extensions
+	for (unsigned i = 0; i < ext_count; ++i) {
+		if (find_extensions(avail_ext_props, avail_extc, &exts[i], 1)) {
+			wlr_log(WLR_DEBUG, "vulkan device extension %s not found",
+				exts[i]);
+			continue;
+		}
+
+		dev->extensions[dev->extension_count++] = exts[i];
+	}
+
+	// for dmabuf importing we require at least the
+	// 'external_memory_fd, external_memory_dma_buf, queue_family_foreign'
+	// extensions. So only enable them if all three are available (assumption
+	// throughout the codebase).
+	const char *names[] = {
+		VK_KHR_EXTERNAL_MEMORY_FD_EXTENSION_NAME,
+		VK_KHR_IMAGE_FORMAT_LIST_EXTENSION_NAME, // or vulkan 1.2
+		VK_EXT_EXTERNAL_MEMORY_DMA_BUF_EXTENSION_NAME,
+		VK_EXT_QUEUE_FAMILY_FOREIGN_EXTENSION_NAME,
+		VK_EXT_IMAGE_DRM_FORMAT_MODIFIER_EXTENSION_NAME,
+	};
+
+	unsigned nc = sizeof(names) / sizeof(names[0]);
+	const char *not_found = find_extensions(avail_ext_props, avail_extc, names, nc);
+	if (not_found) {
+		wlr_log(WLR_ERROR, "vulkan: required device extension %s not found",
+			not_found);
+		goto error;
+	}
+
+	for (unsigned i = 0u; i < nc; ++i) {
+		dev->extensions[dev->extension_count++] = names[i];
+	}
+
+	// queue families
+	{
+		uint32_t qfam_count;
+		vkGetPhysicalDeviceQueueFamilyProperties(phdev, &qfam_count, NULL);
+		assert(qfam_count > 0);
+		VkQueueFamilyProperties queue_props[qfam_count];
+		vkGetPhysicalDeviceQueueFamilyProperties(phdev, &qfam_count,
+			queue_props);
+
+		bool graphics_found = false;
+		for (unsigned i = 0u; i < qfam_count; ++i) {
+			graphics_found = queue_props[i].queueFlags & VK_QUEUE_GRAPHICS_BIT;
```

Comment:
This needs quite a bit more logic to decide which queue to use.
Consider needing 2 queue families - one for uploads, and one for compute. It's always faster to use a transfer queue for uploads, and always faster to use compute queues for compute operations. If there's no dedicated transfer queue, it's always faster to use the dedicated compute queue for transfers. If any of them are unavailable, the graphics queue is guaranteed to always support both the compute and transfer functionality.

Reply:
> It's always faster to use a transfer queue for uploads, and always faster to use compute queues for compute operations.

This is not true. Transfer queues are often *slower* than using the graphics queue. The only time they're a gain is if it frees up the graphics queue to do other work in the meantime. The advantage of the transfer queue is parallelism, not performance.

The same is true for compute queues. Unless you're gaining serious parallelism from doing so, using more than one queue is not a gain.

Reply:
Fair enough, drivers do a lot behind the scenes. Nvidia for example expose 8 or so graphics queues, but they're really just time-slicing a single queue IIRC.
I think I do remember a benchmark you posted last year showing that graphics queues are faster at transfers than transfer queues.
Still, if the situation is appropriate, such as you've submitted an operation to composite and present a finished image, and before that has finished, you get a frame that needs uploading, it would be worth to use a separate transfer or graphics or compute queue. And this should happen in the majority (99%) of cases, since there's really no point in waiting for an image to be presented to process what's available for next frame.